### PR TITLE
VR-8091 LazyList.find should receive *args instead of a list of parameters.

### PR DESCRIPTION
**Pull request by nhatsmrt**
Let me know what y'all think about this change.

### CODE REVIEW

**Comment by nhatsmrt** [comment]
@convoliution I don't actually know how to document variadic functions, so I'm just winging it.

File: client/verta/verta/_internal_utils/_utils.py [hunk 1]
```diff
@@ -260,19 +260,19 @@ def __len__(self):
...
+        A predicate is a string containing a simple boolean expression consisting of:
 
             - a dot-delimited property such as ``metrics.accuracy``
             - a Python boolean operator such as ``>=``
             - a literal value such as ``.8``
 
         Parameters
         ----------
-        where : str or list of str
+        args : strs or a list of str
```
Comment (convoliution) [inline client/verta/verta/_internal_utils/_utils.py]:
```suggestion
        *args : strs
```

Comment (convoliution) [reply client/verta/verta/_internal_utils/_utils.py]: Even though it takes a list, I think we should have an official preference that the inputs be standalone `str`s

Comment (nhatsmrt) [reply client/verta/verta/_internal_utils/_utils.py]: Makes sense to me.

Comment (nhatsmrt) [reply client/verta/verta/_internal_utils/_utils.py]: Fixed.

File: client/verta/verta/_internal_utils/_utils.py [hunk 2]
```diff
@@ -283,16 +283,21 @@ def find(self, where):
...
+            # alternatively:
             runs.find(["hyperparameters.hidden_size == 256",
                        "metrics.accuracy >= .8"])
             # <ExperimentRuns containing 3 runs>
 
         """
-        new_list = copy.deepcopy(self)
+        if len(args) == 1 and isinstance(args[0], list):
+            # to keep backward compatibility, in case user pass in a list
+            return self.find(*args[0])
```
Comment (convoliution) [inline client/verta/verta/_internal_utils/_utils.py:300]: To be extra extra cautious, I think we should also add an `elif` that checks that all `args` are `str`
```
elif not all():  # isinstance(, six.string_types)
    raise TypeError
```

**Review by convoliution — CHANGES_REQUESTED** [review]
One last suggestion

File: client/verta/verta/_internal_utils/_utils.py (hunk 2 above)
Comment (nhatsmrt) [reply client/verta/verta/_internal_utils/_utils.py:300]: Typecheck added. Also add a check in the test to verify new behavior.

File: client/verta/tests/test_entities.py [hunk 3]
```diff
@@ -604,6 +604,17 @@ def test_find_by_tag(self, client):
...
+        runs_with_tag[-1].log_hyperparameter("some-hyper", 1)
+        # compound conditions:
+        assert len(expt.expt_runs.find(["tags ~= {}".format(tag), "hyperparameters.some-hyper == 1"])) == 1  # old syntax
+        assert len(expt.expt_runs.find("tags ~= {}".format(tag), "hyperparameters.some-hyper == 1")) == 1  # new syntax
+
+        # if any predicate is not string, should fail:
+        with pytest.raises(TypeError) as excinfo:
+            expt.expt_runs.find("tag ~= {}".format(tag), 1234)
+
+        assert "predicates must all be strings" in str(excinfo.value).strip()
```
Comment (convoliution) [inline client/verta/tests/test_entities.py:616]: Learned this little trick from Evgenii!
```suggestion
        with pytest.raises(TypeError, match="predicates must all be strings"):
            expt.expt_runs.find("tag ~= {}".format(tag), 1234)
```

File: client/verta/verta/_internal_utils/_utils.py [hunk 4]
```diff
@@ -283,16 +283,23 @@ def find(self, where):
...
+        elif not all([isinstance(predicate, six.string_types) for predicate in args]):
```
Comment (convoliution) [inline client/verta/verta/_internal_utils/_utils.py]: Listification isn't necessary for `all()`
```suggestion
        elif not all(isinstance(predicate, six.string_types) for predicate in args):
```

File: client/verta/tests/test_entities.py (hunk 3 above)
Comment (convoliution) [inline client/verta/tests/test_entities.py]: Learned this little trick from Evgenii!
```suggestion
        with pytest.raises(TypeError, match="predicates must all be strings"):
            expt.expt_runs.find("tag ~= {}".format(tag), 1234)
```

**Review by convoliution — APPROVED** [review]
Two minor suggestions; looks good to me!

File: client/verta/tests/test_entities.py (hunk 3 above)
Comment (nhatsmrt) [reply client/verta/tests/test_entities.py]: Nice! Fixed.

File: client/verta/verta/_internal_utils/_utils.py (hunk 4 above)
Comment (nhatsmrt) [reply client/verta/verta/_internal_utils/_utils.py]: Fixed.

**Pull request merged**